Pipeline aborts in <40s: `git worktree add` races `.git/config.lock` against state-store git ops (regression after #1863)

## Symptom

Two consecutive SDLC pipelines on `issue-2261` failed identically inside the refine phase, both within ~37s of submission:

| Pipeline | Submitted | `pipeline.failed` |
|---|---|---|
| `issue-2261` | 2026-04-30 01:54:28 UTC | 2026-04-30 01:55:03 UTC (35s) |
| `issue-2261-retry` | 2026-04-30 02:00:04 UTC | 2026-04-30 02:00:41 UTC (37s) |

The operator-visible signal was an `OVERSEER_ALERT` ("Overseer container was respawned. Old container … exited with code 1.") followed ~5s later by `pipeline.failed`, with all three BRC agents (refiner, reviewer_refine, reviewer_agent_design) stuck in `producer_phase: WORKING` / `reviewer_phase: WORKING`, none confirmed.

The overseer respawn alert was misleading — the overseer was a victim, not a cause. Orchestrator killed the overseer when it aborted the phase.

## Proximate root cause

One of the four parallel `git worktree add` invocations the gateway runs at agent-spawn time loses a race on the bare repo's `.git/config.lock`. From `kubectl logs gateway` for `issue-2261-retry` (verbatim):

```
2026-04-30 02:00:12 [ERROR] gateway: worktree_create per-repo failure
  repo_name=egg container_id=issue-2261-retry-reviewer_refine
  base_branch=origin/main assigned_branch=egg/issue-2261-retry
  error_type=RuntimeError [/app/gateway.py:7568]
Traceback (most recent call last):
  File "/app/gateway.py", line 7555, in worktree_create
    info = manager.create_worktree(repo_name=repo_name, ...)
  File "/app/worktree_manager.py", line 467, in create_worktree
    raise RuntimeError(f"Failed to create worktree: {result.stderr}")
RuntimeError: Failed to create worktree: Preparing worktree
  (new branch 'egg/issue-2261-retry-reviewer_refine/work')
error: could not lock config file .git/config: File exists
error: unable to write upstream branch configuration
hint: After fixing the error cause you may try to fix up
hint: the remote tracking information by invoking:
hint:   git branch --set-upstream-to=origin/refs/heads/main
```

`reviewer_refine` therefore starts without a worktree. The BRC concurrent_executor sees the missing agent, accumulates `recent_failures` from the resulting cascade (the other 3 agents' own gateway operations also stumble), and aborts the phase at `concurrent_executor.py:595`:

```
2026-04-30 02:00:47 [ERROR] orchestrator.concurrent_executor:
  Multiple agent failures, phase aborted
  recent_failures=2 error="Container exited with code 1"
  pipeline_id=issue-2261-retry [/app/concurrent_executor.py:595]
```

The phase abort terminates all four containers (3 BRC + overseer). The OVERSEER_ALERT we see in operator-visible streams is the orchestrator's auto-respawn briefly cycling the overseer pod between phase-abort decision and full teardown.

## Why the per-repo lock from #1863 doesn't catch this

#1857 / PR #1863 added `_get_repo_lock(repo_name)` (a `threading.Lock`) in `gateway/worktree_manager.py`, which serializes worktree creates **inside the gateway process**. That fix is intact — concurrent gateway requests are now serialized.

The race that's fatal today is **between the gateway and the orchestrator's `state_store`**, both of which run git against the same shared bare repo at `/home/egg/repos/egg/.git`. They are different processes (different k8s pods); a `threading.Lock` cannot synchronize them.

Evidence of the cross-process pressure on `.git/config`. Excerpts from `kubectl logs gateway`, agent-spawn window for `issue-2261-retry`:

```
02:00:04  push egg/pipeline-state                  (state_store)
02:00:06  Worktree created issue-2261-retry        (bootstrap)
02:00:06  push egg/pipeline-state                  (state_store)
02:00:08  push HEAD:refs/heads/egg/issue-2261-retry
02:00:09  push egg/pipeline-state                  (state_store)
02:00:10  push egg/pipeline-state                  (state_store)
02:00:10  Worktree created issue-2261-retry-overseer
02:00:11  Worktree created issue-2261-retry-refiner
02:00:12  Worktree CREATE FAILED issue-2261-retry-reviewer_refine
          (".git/config: File exists" lock race)
02:00:12  Worktree created issue-2261-retry-reviewer_agent_design
```

Five `egg/pipeline-state` pushes interleave with four `git worktree add` invocations inside 6 seconds. The state-store's commits (writing `pipeline state JSONs into a worktree at /home/egg/.egg-state/pipeline-worktree-egg`) and pushes (executed by the gateway via the launcher-secret authenticated path) both touch `.git/config` directly or trigger `.git/config.lock` acquisition. One of them holds the lock while the gateway's `git worktree add -b egg/issue-2261-retry-reviewer_refine/work` tries to write `branch.<name>.remote` into `.git/config`, and the lock contention raises.

## Compounding orchestrator-side failure (separate bug, observed in same logs)

`orchestrator/state_store.py:_run_git` is itself failing on every commit attempt during this window with a different error — its administrative gitdir path is missing:

```
2026-04-30 02:00:47 [ERROR] state_store.py:714  Failed to commit pipeline state
  for issue-2261-retry; file is saved on disk, commit will be retried on next save

GitOperationError: Git command failed:
  fatal: not a git repository:
  /home/egg/repos/egg/.git/worktrees/pipeline-worktree-egg

CalledProcessError: Command
  '['git', '-c', 'core.hooksPath=/dev/null',
    '-C', '/home/egg/.egg-state/pipeline-worktree-egg',
    'add', '.egg-state/pipelines/issue-2261-retry.json']'
  returned non-zero exit status 128.
```

The `pipeline-worktree-egg` worktree's `.git` file points at `/home/egg/repos/egg/.git/worktrees/pipeline-worktree-egg`, but that administrative directory does not exist — the worktree's gitdir was orphaned at some point. `state_store.py:284` later logs `Worktree validation failed` warnings on every retry. This blocks pipeline-state commits but does not by itself fail the pipeline (state survives on disk per the log message). Its self-heal in #2140 / #2167 didn't fire here.

This may share a root cause with the worktree-create race (the same `.git/config.lock` contention or a previous abnormal cleanup pruning the gitdir entry). Worth investigating jointly.

## Reproduction

Submit any SDLC pipeline that triggers the full BRC trio (3 reviewers + producer + overseer) under the current orchestrator/gateway revisions running in `egg-system` (orchestrator pod `orchestrator-58d7f9b68d-v42lc`, gateway pod `gateway-5d594779fc-bpc2w`, both started 2026-04-30 ~01:56 UTC). Two consecutive submissions failed with this signature; reproduction rate is high during pipeline-state-commit-heavy windows.

## Acceptance criteria

- A pipeline that spawns 4+ containers concurrently while state-store is committing pipeline JSON does not lose a `.git/config.lock` race against the gateway's `git worktree add`.
- The fix serializes git operations against the shared bare repo across processes (e.g. `flock(2)` on a sentinel file in `.git/`, or routing all bare-repo writes through a single broker process), not just within the gateway thread pool.
- Existing #1857 / #1863 in-process serialization is preserved.
- Add a regression test that exercises N-way concurrent worktree creates on the same bare repo while state-store commits run in a sibling process; assert no `.git/config.lock` failures.

## Out of scope (file separately if needed)

- The orphaned `pipeline-worktree-egg` gitdir issue noted above — almost certainly related, but distinct enough to fix independently.
- The recurring `git_execute_blocked` warnings for agents running `git log --oneline -n 5` — gateway allowlist rejects `-n`. Not fatal here, just noise, but ~10 occurrences during this 35s window suggest it should be triaged.

## Related

- **#1857** — original concurrent worktree-create race (gateway-internal). Closed by **#1863**.
- **#2140** / **#2151** — make state-store `_ensure_worktree` idempotent for broken worktrees. Did not fire here.
- **#2167** — state-store self-heal. Did not fire here.
- **#2235** — stop ENOENT race + mark-FAILED crash. Sibling concurrency fix.
- **#2238** — serialize `_ensure_worktree` to close concurrent-recovery race. Same neighborhood.
- **#1842** — partial-spawn-failure aborts whole phase. Tangential but relevant.

## Logs

`mcp__egg__get_service_logs` for both `orchestrator` and `gateway` covering ~02:00:00–02:00:50 UTC has the full traceback chain. Available on operator's host at:

- `~/.claude/projects/.../tool-results/mcp-egg-get_service_logs-1777515900839.txt` (orchestrator)
- `~/.claude/projects/.../tool-results/mcp-egg-get_service_logs-1777515901234.txt` (gateway)

--- a/gateway/Dockerfile
+++ b/gateway/Dockerfile
@@ -66,12 +66,13 @@ WORKDIR /app
 # Copy gateway code
 COPY gateway/*.py ./
 
-# Copy shared modules (egg_logging for logging, egg_config for config parsing, egg_contracts for contract API, egg_restrictions for agent role enforcement, egg_health for health tracking)
+# Copy shared modules (egg_logging for logging, egg_config for config parsing, egg_contracts for contract API, egg_restrictions for agent role enforcement, egg_health for health tracking, egg_git for cross-process git locking)
 COPY shared/egg_logging/ ./egg_logging/
 COPY shared/egg_config/ ./egg_config/
 COPY shared/egg_contracts/ ./egg_contracts/
 COPY shared/egg_restrictions/ ./egg_restrictions/
 COPY shared/egg_health/ ./egg_health/
+COPY shared/egg_git/ ./egg_git/
 
 # Copy config module (repo_config needed by github_client for user mode)
 # Copy to both /config/ (for package-style imports) and /app/ (for PYTHONPATH)

--- a/gateway/tests/test_worktree_manager.py
+++ b/gateway/tests/test_worktree_manager.py
@@ -1338,6 +1338,119 @@ class TestWorktreeManagerConcurrency:
             assert git_file.is_file(), f"{cid}: .git should be a file"
             assert git_file.read_text().strip().startswith("gitdir:")
 
+    def test_concurrent_create_worktree_with_sibling_state_writes(self, git_repo):
+        """Concurrent worktree creates do not lose ``.git/config`` lock races to
+        a sibling process running state-store-style git operations on the
+        same bare repo (#2311).
+
+        The pre-#2311 gateway only serialised worktree creates inside its
+        own process; a separate process (the orchestrator's state-store)
+        could write to ``.git/config`` between the gateway's claim and
+        commit of ``.git/config.lock``, and ``git worktree add`` would
+        fail with ``could not lock config file .git/config: File exists``.
+        With the cross-process flock, both sides cooperatively serialise
+        on a shared sentinel inside ``.git/`` — ``git worktree add`` no
+        longer races.
+        """
+        import os as _os
+        import subprocess as _sp
+        import sys as _sys
+        import textwrap
+        import threading
+
+        worktree_base, repos_base, repo_dir = git_repo
+        manager = WorktreeManager(worktree_base=worktree_base, repos_base=repos_base)
+
+        # Sibling subprocess: bursts of ``git config`` writes guarded by
+        # ``bare_repo_lock`` — i.e. the same lock the gateway acquires.
+        # Each write goes through ``.git/config.lock``, the same file that
+        # ``git worktree add`` claims when it sets ``branch.<x>.remote``.
+        ready_sentinel = repo_dir / ".contender-ready"
+        contender_script = textwrap.dedent(
+            f"""
+            import subprocess, sys, time
+            sys.path.insert(0, {str(Path(__file__).resolve().parent.parent.parent / "shared")!r})
+            from egg_git.cross_process_lock import bare_repo_lock
+
+            repo = {str(repo_dir)!r}
+            open({str(ready_sentinel)!r}, "w").close()
+            deadline = time.monotonic() + 8
+            i = 0
+            while time.monotonic() < deadline:
+                with bare_repo_lock(repo):
+                    subprocess.run(
+                        ["git", "-C", repo, "config",
+                         f"egg.statetest{{i}}", str(i)],
+                        check=False, capture_output=True,
+                    )
+                i += 1
+            """
+        )
+        contender = _sp.Popen(
+            [_sys.executable, "-c", contender_script],
+            env={**_os.environ, "PYTHONUNBUFFERED": "1"},
+            stdout=_sp.PIPE,
+            stderr=_sp.PIPE,
+        )
+
+        try:
+            # Wait for the contender to start running before launching
+            # threads so the race window actually overlaps.
+            import time as _time
+
+            deadline = _time.monotonic() + 5
+            while not ready_sentinel.exists() and _time.monotonic() < deadline:
+                if contender.poll() is not None:
+                    out, err = contender.communicate(timeout=1)
+                    pytest.fail(
+                        f"contender exited early rc={contender.returncode}: "
+                        f"stdout={out!r} stderr={err!r}"
+                    )
+                _time.sleep(0.01)
+            assert ready_sentinel.exists(), "contender did not signal readiness"
+
+            results: dict[str, object] = {}
+            container_ids = [f"container-{i}" for i in range(6)]
+            barrier = threading.Barrier(len(container_ids))
+
+            def create(cid: str) -> None:
+                try:
+                    barrier.wait(timeout=10)
+                    # ``assigned_branch`` triggers ``_configure_push_upstream``,
+                    # which writes ``branch.<x>.remote`` / ``branch.<x>.merge``
+                    # into ``.git/config``.  That is the same write that races
+                    # ``.git/config.lock`` in #2311.
+                    results[cid] = manager.create_worktree(
+                        "test-repo",
+                        cid,
+                        assigned_branch="egg/issue-2311-target",
+                    )
+                except Exception as exc:
+                    results[cid] = exc
+
+            threads = [threading.Thread(target=create, args=(cid,)) for cid in container_ids]
+            for t in threads:
+                t.start()
+            for t in threads:
+                t.join(timeout=60)
+        finally:
+            contender.terminate()
+            try:
+                contender.wait(timeout=5)
+            except _sp.TimeoutExpired:
+                contender.kill()
+                contender.wait(timeout=2)
+
+        for cid in container_ids:
+            assert cid in results, f"Thread for {cid} did not finish"
+            info = results[cid]
+            if isinstance(info, Exception):
+                # Surface the exact error message — the regression
+                # signature is ``could not lock config file .git/config``.
+                pytest.fail(f"{cid} failed: {info!r}")
+            assert isinstance(info, WorktreeInfo)
+            assert info.worktree_path.exists()
+
 
 class TestRunGitWorktreeAddRetry:
     """Tests for _run_git_worktree_add retry logic on index.lock contention."""

--- a/gateway/tests/test_worktree_manager.py
+++ b/gateway/tests/test_worktree_manager.py
@@ -1377,10 +1377,12 @@ class TestWorktreeManagerConcurrency:
             deadline = time.monotonic() + 8
             i = 0
             while time.monotonic() < deadline:
+                # Reuse a single key — same race surface on
+                # ``.git/config.lock`` without bloating ``.git/config``
+                # with thousands of distinct keys over the test window.
                 with bare_repo_lock(repo):
                     subprocess.run(
-                        ["git", "-C", repo, "config",
-                         f"egg.statetest{{i}}", str(i)],
+                        ["git", "-C", repo, "config", "egg.statetest", str(i)],
                         check=False, capture_output=True,
                     )
                 i += 1

--- a/gateway/worktree_manager.py
+++ b/gateway/worktree_manager.py
@@ -29,7 +29,9 @@ _shared_path = Path(__file__).parent.parent.parent / "shared"
 if _shared_path.exists():
     sys.path.insert(0, str(_shared_path))
 import contextlib
+from collections.abc import Generator
 
+from egg_git.cross_process_lock import bare_repo_lock
 from egg_logging import get_logger
 
 # Import git_cmd helper
@@ -319,23 +321,26 @@ class WorktreeManager:
             # Ensure ownership is correct (may have been created with different uid/gid)
             self._chown_recursive(worktree_path, uid, gid)
             self._chown_single(worktree_path.parent, uid, gid)
-            # Re-apply push upstream config in case the assigned branch
-            # changed across calls (idempotent when unchanged).
-            self._configure_push_upstream(main_repo, branch_name, assigned_branch)
-            # Reset HEAD to a known-good ref so we don't inherit a stale
-            # left-over HEAD from a prior pipeline that collided on
-            # ``container_id`` (deterministic pipeline_id can collide when
-            # the same issue is resubmitted — see #2222).  Without this,
-            # the new pipeline's first push hits non-fast-forward and
-            # the reconcile path can absorb upstream main commits onto
-            # the pipeline branch.
-            self._reset_reused_worktree_to_safe_ref(
-                worktree_path=worktree_path,
-                main_repo=main_repo,
-                container_id=container_id,
-                assigned_branch=assigned_branch,
-                base_branch=base_branch,
-            )
+            # Re-apply push upstream config and reset to a safe ref under
+            # the per-repo lock — both write to ``.git/config`` (and the
+            # latter to ``.git/index``), and concurrent callers without
+            # the lock race on ``.git/config.lock`` (#2311).
+            with self._get_repo_lock(repo_name):
+                self._configure_push_upstream(main_repo, branch_name, assigned_branch)
+                # Reset HEAD to a known-good ref so we don't inherit a stale
+                # left-over HEAD from a prior pipeline that collided on
+                # ``container_id`` (deterministic pipeline_id can collide when
+                # the same issue is resubmitted — see #2222).  Without this,
+                # the new pipeline's first push hits non-fast-forward and
+                # the reconcile path can absorb upstream main commits onto
+                # the pipeline branch.
+                self._reset_reused_worktree_to_safe_ref(
+                    worktree_path=worktree_path,
+                    main_repo=main_repo,
+                    container_id=container_id,
+                    assigned_branch=assigned_branch,
+                    base_branch=base_branch,
+                )
             # Return info about existing worktree
             return WorktreeInfo(
                 container_id=container_id,
@@ -496,8 +501,11 @@ class WorktreeManager:
         # Point the per-worktree local branch at the assigned remote branch
         # so `git push` resolves to a refspec the gateway will accept
         # (#1809).  Must happen before returning so the sandbox's push
-        # client sees the config on its first push.
-        self._configure_push_upstream(main_repo, branch_name, assigned_branch)
+        # client sees the config on its first push.  Held under the
+        # per-repo lock so the ``.git/config`` write does not race the
+        # state-store's commits in the orchestrator pod (#2311).
+        with self._get_repo_lock(repo_name):
+            self._configure_push_upstream(main_repo, branch_name, assigned_branch)
 
         # Find the actual git dir (git names it based on worktree basename)
         git_dir = self._find_worktree_git_dir(main_repo, worktree_path)
@@ -584,12 +592,27 @@ class WorktreeManager:
             git_dir=self._find_worktree_git_dir(main_repo, worktree_path),
         )
 
-    def _get_repo_lock(self, repo_name: str) -> threading.Lock:
-        """Get or create a per-repo lock for serializing git operations."""
+    @contextlib.contextmanager
+    def _get_repo_lock(self, repo_name: str) -> Generator[None]:
+        """Hold the per-repo lock for serializing git operations.
+
+        Combines two layers:
+
+        * A per-repo ``threading.Lock`` for in-process serialization (the
+          original #1857 / #1863 fix).
+        * ``bare_repo_lock`` for cross-process serialization against the
+          orchestrator's state-store, which runs git from a different pod
+          but shares the same hostPath-mounted bare repo (#2311).  Without
+          this, parallel ``git worktree add`` calls race the state-store's
+          commits on ``.git/config.lock`` and fail with ``could not lock
+          config file .git/config: File exists``.
+        """
         with self._repo_locks_guard:
             if repo_name not in self._repo_locks:
                 self._repo_locks[repo_name] = threading.Lock()
-            return self._repo_locks[repo_name]
+            thread_lock = self._repo_locks[repo_name]
+        with thread_lock, bare_repo_lock(self.repos_base / repo_name):
+            yield
 
     def _configure_push_upstream(
         self,

--- a/gateway/worktree_manager.py
+++ b/gateway/worktree_manager.py
@@ -693,7 +693,11 @@ class WorktreeManager:
         """
         # Best-effort fetch so the remote-tracking refs are current; if
         # this fails we still attempt the reset against whatever local
-        # state we have.
+        # state we have.  This runs inside the cross-process lock (the
+        # caller holds it across this whole method) so the timeout
+        # caps how long every other state-store commit / worktree
+        # create on this repo can be blocked by a slow remote — keep
+        # it tight.
         try:
             subprocess.run(
                 git_cmd("fetch", "origin"),
@@ -701,7 +705,7 @@ class WorktreeManager:
                 capture_output=True,
                 text=True,
                 check=False,
-                timeout=120,
+                timeout=30,
             )
         except (subprocess.TimeoutExpired, OSError) as exc:
             logger.warning(

--- a/orchestrator/state_store.py
+++ b/orchestrator/state_store.py
@@ -29,6 +29,7 @@ from datetime import UTC, datetime
 from pathlib import Path
 from typing import Any, ClassVar, Literal
 
+from egg_git.cross_process_lock import lock_path_for_repo
 from models import Pipeline, PipelineMode, PipelinePhase, PipelineStatus
 from pydantic import ValidationError
 
@@ -153,8 +154,17 @@ class StateStore:
 
     @property
     def _lock_path(self) -> Path:
-        """Lock file for cross-process git serialization."""
-        return self._worktree_dir.parent / ".git-ops.lock"
+        """Lock file for cross-process git serialization.
+
+        Located inside the bare repo's ``.git/`` directory so the same
+        sentinel inode is visible to the gateway pod, which mounts the
+        repo from the same hostPath.  Until #2311, the lock lived under
+        ``self._worktree_dir.parent`` — a pod-local ``emptyDir`` — so the
+        gateway and orchestrator never actually synchronised, and a
+        ``git worktree add`` could race a state-store commit on
+        ``.git/config.lock`` and fail.
+        """
+        return lock_path_for_repo(self.repo_path)
 
     @classmethod
     def _get_flock_fd(cls, lock_path: Path) -> int:

--- a/orchestrator/tests/test_state_store.py
+++ b/orchestrator/tests/test_state_store.py
@@ -1109,7 +1109,10 @@ class TestRunGitLocking:
         store = StateStore(tmp_path, worktree_dir=worktree_dir)
         store._worktree = worktree_dir
 
-        lock_file = tmp_path / ".git-ops.lock"
+        # Lock file lives inside the bare repo's .git/ so the gateway pod
+        # (which mounts /home/egg/repos/ from the same hostPath) sees the
+        # same inode and flock serialises both processes (#2311).
+        lock_file = tmp_path / ".git" / ".egg-cross-process.lock"
         assert not lock_file.exists()
 
         with store._git_op():

--- a/shared/egg_git/__init__.py
+++ b/shared/egg_git/__init__.py
@@ -4,8 +4,11 @@ Git utilities for egg.
 Provides common git operations that can be reused across processors and tasks.
 """
 
+from .cross_process_lock import bare_repo_lock, lock_path_for_repo
 from .default_branch import get_default_branch
 
 __all__ = [
+    "bare_repo_lock",
     "get_default_branch",
+    "lock_path_for_repo",
 ]

--- a/shared/egg_git/cross_process_lock.py
+++ b/shared/egg_git/cross_process_lock.py
@@ -1,0 +1,116 @@
+"""Cross-process serialization for git operations on a shared bare repo.
+
+The gateway pod and orchestrator pod both run git against the same shared
+bare repo at ``/home/egg/repos/<repo>/.git`` (the same hostPath is mounted
+into each pod).  A ``threading.Lock`` only synchronises calls inside one
+process — nothing stops the gateway's ``git worktree add`` from racing the
+orchestrator's state-store commit on ``.git/config.lock`` (#2311).
+
+This module wraps that file with an ``fcntl.flock`` so both processes
+serialise on the same inode.  The lock file lives at
+``<repo_path>/.git/.egg-cross-process.lock`` — inside the resource being
+protected, on the shared mount, so any process that can run git against
+the repo can also see and acquire the lock.
+
+Reentrancy: nested ``with bare_repo_lock(repo)`` blocks in the same
+process do not block — the inner acquisitions just bump a depth counter.
+This matches the orchestrator state-store's existing pattern, where a
+compound operation (e.g. ``add → commit``) holds the lock across several
+inner ``_run_git`` calls.
+"""
+
+from __future__ import annotations
+
+import contextlib
+import fcntl
+import os
+import threading
+from collections.abc import Generator
+from pathlib import Path
+from typing import Final
+
+LOCK_FILENAME: Final = ".egg-cross-process.lock"
+
+
+def lock_path_for_repo(repo_path: Path | str) -> Path:
+    """Return the cross-process lock path for ``repo_path``.
+
+    The path is ``<repo_path>/.git/<LOCK_FILENAME>``.  ``repo_path`` must
+    be the main repo (with ``.git`` as a directory).  Worktree paths
+    (where ``.git`` is a file) are not valid arguments — callers should
+    resolve to the main repo first.
+    """
+    return Path(repo_path) / ".git" / LOCK_FILENAME
+
+
+class _RepoLockState:
+    __slots__ = ("rlock", "fd", "depth")
+
+    def __init__(self, fd: int) -> None:
+        self.rlock = threading.RLock()
+        self.fd = fd
+        # Nesting depth, only mutated while ``rlock`` is held.
+        self.depth = 0
+
+
+# Per-process state, keyed by stringified repo path.  Each entry owns a
+# long-lived file descriptor on the lock file plus a reentrant thread
+# lock so multiple threads in this process queue locally rather than all
+# bouncing on the kernel's flock queue.
+_state_lock = threading.Lock()
+_per_repo_state: dict[str, _RepoLockState] = {}
+
+
+def _get_state(repo_path: Path) -> _RepoLockState:
+    key = str(repo_path)
+    with _state_lock:
+        state = _per_repo_state.get(key)
+        if state is None:
+            lock_path = lock_path_for_repo(repo_path)
+            lock_path.parent.mkdir(parents=True, exist_ok=True)
+            fd = os.open(str(lock_path), os.O_CREAT | os.O_RDWR, 0o644)
+            state = _RepoLockState(fd=fd)
+            _per_repo_state[key] = state
+        return state
+
+
+@contextlib.contextmanager
+def bare_repo_lock(repo_path: Path | str) -> Generator[None]:
+    """Hold the cross-process bare-repo lock for ``repo_path``.
+
+    Acquires, in order, a reentrant in-process thread lock and an
+    ``fcntl.flock`` ``LOCK_EX`` on a sentinel file under ``.git/``.
+    Releases both on exit.  Reentrant within a single process via a
+    depth counter — only the outermost caller actually issues the
+    ``flock`` syscall.
+    """
+    state = _get_state(Path(repo_path))
+    state.rlock.acquire()
+    try:
+        if state.depth == 0:
+            fcntl.flock(state.fd, fcntl.LOCK_EX)
+        state.depth += 1
+        try:
+            yield
+        finally:
+            state.depth -= 1
+            if state.depth == 0:
+                fcntl.flock(state.fd, fcntl.LOCK_UN)
+    finally:
+        state.rlock.release()
+
+
+def reset_for_tests() -> None:
+    """Drop the per-repo state cache.  Test-only.
+
+    Closes every cached file descriptor and clears the registry so a
+    fresh test run starts from a clean slate.  Calling this while a
+    lock is held in another thread is undefined behaviour.
+    """
+    with _state_lock:
+        for state in _per_repo_state.values():
+            try:
+                os.close(state.fd)
+            except OSError:
+                pass
+        _per_repo_state.clear()

--- a/shared/egg_git/cross_process_lock.py
+++ b/shared/egg_git/cross_process_lock.py
@@ -62,7 +62,15 @@ _per_repo_state: dict[str, _RepoLockState] = {}
 
 
 def _get_state(repo_path: Path) -> _RepoLockState:
-    key = str(repo_path)
+    # Resolve so two equivalent path forms (abs vs rel, with/without
+    # trailing slash, symlinks) hit the same cache entry.  flock keys
+    # on the inode regardless, but a single fd avoids redundant state
+    # and a self-deadlock risk if a future caller relies on RLock
+    # reentrancy across forms.
+    try:
+        key = str(repo_path.resolve())
+    except OSError:
+        key = str(repo_path)
     with _state_lock:
         state = _per_repo_state.get(key)
         if state is None:

--- a/shared/tests/test_cross_process_lock.py
+++ b/shared/tests/test_cross_process_lock.py
@@ -96,7 +96,7 @@ def test_serialises_across_processes(tmp_path: Path) -> None:
         with bare_repo_lock({str(tmp_path)!r}):
             # Signal "lock held" by writing to a sentinel file.
             open({str(tmp_path / "held")!r}, "w").close()
-            time.sleep(0.5)
+            time.sleep(1.0)
         """
     )
 
@@ -121,7 +121,8 @@ def test_serialises_across_processes(tmp_path: Path) -> None:
             proc.kill()
             proc.wait(timeout=2)
 
-    # The child held the lock for ~0.5s after writing the sentinel; the
-    # parent should have waited at least most of that time.  Allow a
-    # generous slack for scheduler jitter on busy CI hosts.
-    assert wait > 0.2, f"parent acquired lock too quickly ({wait:.3f}s) — flock did not block"
+    # The child held the lock for ~1.0s after writing the sentinel; the
+    # parent should have waited at least half that time.  The 0.5s slack
+    # absorbs the parent's sentinel-detection latency (the polling loop
+    # ticks every 10ms) plus scheduler jitter on busy CI hosts.
+    assert wait > 0.5, f"parent acquired lock too quickly ({wait:.3f}s) — flock did not block"

--- a/shared/tests/test_cross_process_lock.py
+++ b/shared/tests/test_cross_process_lock.py
@@ -1,0 +1,127 @@
+"""Tests for ``egg_git.cross_process_lock``.
+
+Covers the per-repo flock primitive that synchronises git operations
+between the gateway and orchestrator pods on a shared bare repo (#2311).
+"""
+
+from __future__ import annotations
+
+import os
+import subprocess
+import sys
+import textwrap
+import threading
+import time
+from pathlib import Path
+
+import pytest
+from egg_git.cross_process_lock import (
+    LOCK_FILENAME,
+    bare_repo_lock,
+    lock_path_for_repo,
+    reset_for_tests,
+)
+
+
+@pytest.fixture(autouse=True)
+def _reset_state():
+    yield
+    reset_for_tests()
+
+
+@pytest.fixture
+def repo(tmp_path: Path) -> Path:
+    """A directory with an empty ``.git/`` so the lock file can land."""
+    (tmp_path / ".git").mkdir()
+    return tmp_path
+
+
+def test_lock_path_lives_inside_dot_git(tmp_path: Path) -> None:
+    assert lock_path_for_repo(tmp_path) == tmp_path / ".git" / LOCK_FILENAME
+
+
+def test_acquiring_creates_lock_file(repo: Path) -> None:
+    expected = repo / ".git" / LOCK_FILENAME
+    assert not expected.exists()
+    with bare_repo_lock(repo):
+        assert expected.exists()
+
+
+def test_reentrant_within_process(repo: Path) -> None:
+    # Three nested ``with`` blocks must not deadlock — only the outermost
+    # actually issues an ``flock`` syscall.
+    with bare_repo_lock(repo):
+        with bare_repo_lock(repo):
+            with bare_repo_lock(repo):
+                pass
+
+
+def test_threads_in_one_process_serialise(repo: Path) -> None:
+    in_section: list[bool] = []
+    overlap = threading.Event()
+
+    def hold(duration: float) -> None:
+        with bare_repo_lock(repo):
+            if any(in_section):
+                overlap.set()
+            in_section.append(True)
+            try:
+                time.sleep(duration)
+            finally:
+                in_section.pop()
+
+    threads = [threading.Thread(target=hold, args=(0.05,)) for _ in range(4)]
+    for t in threads:
+        t.start()
+    for t in threads:
+        t.join(timeout=5)
+
+    assert not overlap.is_set(), "threads observed each other inside the lock"
+
+
+def test_serialises_across_processes(tmp_path: Path) -> None:
+    """A subprocess holding the lock blocks the parent until it releases.
+
+    Without flock-on-shared-inode (the #2311 fix), the parent would
+    proceed immediately and the assertion below would fail by a wide
+    margin.
+    """
+    (tmp_path / ".git").mkdir()
+
+    holder_script = textwrap.dedent(
+        f"""
+        import sys, time
+        sys.path.insert(0, {str(Path(__file__).resolve().parent.parent)!r})
+        from egg_git.cross_process_lock import bare_repo_lock
+        with bare_repo_lock({str(tmp_path)!r}):
+            # Signal "lock held" by writing to a sentinel file.
+            open({str(tmp_path / "held")!r}, "w").close()
+            time.sleep(0.5)
+        """
+    )
+
+    proc = subprocess.Popen(
+        [sys.executable, "-c", holder_script],
+        env={**os.environ, "PYTHONUNBUFFERED": "1"},
+    )
+
+    try:
+        # Wait until the child has actually acquired the lock.
+        deadline = time.monotonic() + 5
+        while not (tmp_path / "held").exists() and time.monotonic() < deadline:
+            time.sleep(0.01)
+        assert (tmp_path / "held").exists(), "child never signalled lock acquisition"
+
+        start = time.monotonic()
+        with bare_repo_lock(tmp_path):
+            wait = time.monotonic() - start
+        proc.wait(timeout=5)
+    finally:
+        if proc.poll() is None:
+            proc.kill()
+            proc.wait(timeout=2)
+
+    # The child held the lock for ~0.5s after writing the sentinel; the
+    # parent should have waited at least most of that time.  Allow a
+    # generous slack for scheduler jitter on busy CI hosts.
+    assert wait > 0.2, f"parent acquired lock too quickly ({wait:.3f}s) — flock did not block"

